File: ParkingLotDesign.java
// Time Complexity : for park(): O(log n) where n is the spot object; unpark: O(log n); for getNextAvailable: O(1), for addParkingspot: O(log n) for each of the calls
// Space Complexity : O(n) for maintaining n pairs of Floors and Spots objects

import java.util.PriorityQueue;
class ParkingLot {
    int maxFloors, spotsPerFloor;
    PriorityQueue<ParkingSpot> pq = new PriorityQueue<>((a,b) -> {
        if(a.floor == b.floor) return a.spot-b.spot;
        return a.floor-b.floor;
    });
    public ParkingLot(int maxFloors, int spotsPerFloor) {
        this.maxFloors = maxFloors;
        this.spotsPerFloor = spotsPerFloor;
    }
    public ParkingSpot park() {
        if(pq.isEmpty()) {
            throw new IllegalStateException("Parking lot is full");
        }
        ParkingSpot re = pq.remove();
        return re;
    }
    public ParkingSpot getNextAvailable() {
        return pq.peek();
    }
    public void unpark(int floor,int spot) {
        ParkingSpot newSpot = new ParkingSpot(floor,spot);
        pq.add(newSpot);
    }
    public void addParkingSpot(int floor,int spot) {
        if(floor > maxFloors) {
            throw new IllegalArgumentException("floor limit exceeded");
        }
        if(spot > spotsPerFloor) {
            throw new IllegalArgumentException("spot limit exceeded");
        }
        ParkingSpot newSpot = new ParkingSpot(floor,spot);
        pq.add(newSpot);
    }
}

class ParkingSpot {
    int floor; int spot;
    public ParkingSpot(int floor, int spot){
        this.floor = floor;
        this.spot = spot;
    }
    public int getSpot(){
        return this.spot;
    }
    public int getFloor(){
        return this.floor;
    }
}

class ParkingLotDesign {
    public static void main(String[] args) {
        ParkingLot pl = new ParkingLot(10,20);
        pl.addParkingSpot(1, 1);
 	    pl.addParkingSpot(2, 1);
 		pl.addParkingSpot(3, 1);
 		pl.addParkingSpot(1, 2);
 		pl.addParkingSpot(2, 2);
 		pl.addParkingSpot(3, 2);
		ParkingSpot n = pl.getNextAvailable();
 		System.out.println("Parked at Floor: " + n.getFloor() + ", Slot: " + n.getSpot());
 		pl.park();
 		ParkingSpot n2 = pl.getNextAvailable();
 		System.out.println("Parked at Floor: " + n2.getFloor() + ", Slot: " + n2.getSpot());
 		pl.unpark(1, 1);
 		ParkingSpot n1 = pl.getNextAvailable();
 		System.out.println("Parked at Floor: " + n1.getFloor() + ", Slot: " + n1.getSpot());
    }
} 

File: parkingLotDesign.py
from collections import deque
import heapq
class ParkingSpot:
    
    # initialize the floor and spot
    def __init__(self,floor,spot):
        self.floor = floor
        self.spot = spot
    
    # get the floor
    def getFloor(self):
        return self.floor
    
    # get the spot
    def getSpot(self):
        return self.spot

class ParkingLot:
    
    # initialize totalFloor and totalSlots avaialble in a singleFloor
    def __init__(self,floors,slots):
        self.floors = floors
        self.slots = slots
        
        # initlize a parking grid space
        self.parkingGrid = [[None for col in range(0,self.slots)]for row in range(0,self.floors)]
        
        # initlize parking space
        self.parkingSpace = []
        
        # add the floors and slots to the minHeap
        for floor in range(0,self.floors):
            for slot in range(0,self.slots):
                heapq.heappush(self.parkingSpace,(floor,slot))
    
    # park the vehicle
    def parkVehicle(self):
        '''extract the min from the queue and park'''
        
        # get the spot i.e extract the minfrom the queue
        availableSpot = heapq.heappop(self.parkingSpace)
        
        # create an object of parkingSpot and park the car in the parkingGrid
        parkingSpot = ParkingSpot(availableSpot[0],availableSpot[1])
        self.parkingGrid[availableSpot[0]][availableSpot[1]] = parkingSpot
        
        return vars(self.parkingGrid[availableSpot[0]][availableSpot[1]])
    
    # unpark the vehicle
    def unparkVehicle(self,parkingSpot: ParkingSpot):
        '''provide the spot and pueh to the queue'''
        # parkingSpot will be the obj of ParkingSpot class
        
        # get the floor
        floor = parkingSpot.getFloor()
        
        # get the spot
        spot = parkingSpot.getSpot()
        
        # add to the priorityQueue
        heapq.heappush(self.parkingSpace,(floor,spot))
        
        # set the parkingGrid to None
        self.parkingGrid[floor][spot] = None        
        
        return self.parkingGrid[floor][spot]

parkingLot = ParkingLot(floors=4,slots=10)

# park 10 vehicles
for count in range(0,10):
    parkingLot.parkVehicle()
    
# unpark the vehicles
parkingLot.unparkVehicle(parkingLot.parkingGrid[0][0])
parkingLot.unparkVehicle(parkingLot.parkingGrid[0][8])
print(f"Floor 0 is:\t{parkingLot.parkingGrid[0]}")